controllers/users.js

```diff
@@ -0,0 +1,410 @@
+var async = require('async'),
+    _ = require('underscore'),
+    db = require('../db');
+
+var getType = function(user, admins) {
+  if (user.roles && user.roles.length) {
+    return user.roles[0];
+  }
+  return admins[user.name] ? 'admin' : 'unknown';
+};
+
+var getDoc = function(id, docs) {
+  return _.findWhere(docs, { _id: id });
+};
+
+var getDocID = function(doc) {
+  if (_.isString(doc)) {
+    return doc;
+  }
+  if (_.isObject(doc)) {
+    return doc._id;
+  }
+};
+
+var getAllUserSettings = function(callback) {
+  var opts = {
+    include_docs: true,
+    key: ['user-settings']
+  };
+  db.medic.view('medic', 'doc_by_type', opts, function(err, results) {
+    if (err) {
+      return callback(err);
+    }
+    callback(null, _.map(results.rows, function(row) {
+      return row.doc;
+    }));
+  });
+};
+
+var getAllUsers = function(callback) {
+  db._users.list({include_docs: true}, function(err, results) {
+    if (err) {
+      return callback(err);
+    }
+    callback(null, results.rows);
+  });
+};
+
+var getFacilities = function(callback) {
+  db.medic.view('medic', 'facilities', {include_docs: true}, function(err, results) {
+    if (err) {
+      return callback(err);
+    }
+    callback(null, _.map(results.rows, function(row) {
+      return row.doc;
+    }));
+  });
+};
+
+var isAPlace = function(place) {
+  return [
+    'district_hospital',
+    'health_center',
+    'clinic'
+  ].indexOf(place.type) !== -1;
+};
+
+var getPlace = function(id, callback) {
+  db.medic.get(id, function(err, place) {
+    if (err) {
+      return callback(err);
+    }
+    if (!isAPlace(place)) {
+      return callback('Wrong type, this is not a place.');
+    }
+    callback(null, place);
+  });
+};
+
+var createUser = function(data, response, callback) {
+  response = response || {};
+  var id = createID(data.username),
+      user = getUserUpdates(id, data);
+  db._users.insert(user, id, function(err, body) {
+    if (err) {
+      return callback('Failed to create user. ' + err);
+    }
+    response.user = {
+      id: body.id,
+      rev: body.rev
+    };
+    callback(null, data, response);
+  });
+};
+
+/*
+ * Warning: not doing validation of the contact data against a form yet.  The
+ * form is user defined in settings so being liberal with what gets saved to
+ * the database. Ideally CouchDB could validate a given object against a form
+ * in validate_doc_update.
+ */
+var createContact = function(data, response, callback) {
+  response = response || {};
+  db.medic.insert(data.contact, function(err, body) {
+    if (err) {
+      return callback('Failed to create contact. ' + err);
```

Comment:
This won't work nicely if err is an Error not a string. Maybe just `callback(err)` instead?